webpack.config.ts

```diff
@@ -118,6 +118,7 @@ const config: webpack.Configuration[] = [
       mousetrap: 'require("./node_modules/mousetrap/mousetrap.js")',
       open: 'require("./node_modules/open/index.js")',
       'xterm-addon-fit': 'require("./node_modules/xterm-addon-fit/lib/xterm-addon-fit.js")',
+      'xterm-addon-image': 'require("./node_modules/xterm-addon-image/lib/xterm-addon-image.js")',
```

Comment:
You'll have to require it in https://github.com/vercel/hyper/blob/canary/bin/snapshot-libs.js as well to put it in v8 snapshot.

Reply:
Putting `xterm-addon-image` in the v8 snapshot currently breaks the build, as it uses a reference to `self`, which does not appear to be available in that environment. Unfortunately I have no idea how to fix that. Is it required to put `xterm-addon-image` into the snapshot? I am not too familiar with electron and v8 snapshots, but from what I read in #5941, #6224, it is some form of caching and only reduces startup time, but is not necessary to run the application and use the dependency therein?

Reply:
yeah it's not necessary to put it in snapshot, I will test it out when I'll pick up the xterm v5 pr (hoping for this weekend).

Reply:
Idk anything about that snapshot functionality, would that also deal/preserve wasm state? If it cannot deal with wasm the same way, then it will prolly not work after resume from snapshot?

The initial addon loading on calling the ctor should be okish - I tried to shape most of the heavier stuff to lazy initialization upon first usage (e.g. the base64 wasm decoder will spawn on first chunk to be decoded). There still might be room for improvement, just file me a note if you find unpleasant blocking stuff in startup profiling.

Reply:
i'm also not sure about the wasm stuff with regards to snapshots, it basically parses js and stores the memory state. I will test and see if it works.

Reply:
Still getting the `self is not defined` error when putting `xterm-addon-image@0.4.1` into the snapshot with xterm 5.2. Everything works fine when using the snapshot from the `canary` branch though, this appears to be the only remaining issue.

Reply:
@marvinruder Does it go away, if you replace `self` with `this` in the bundled JS file? And more important, does the addon still work with this change? (In theory there should not be any explicit references to the global context from within the code, but I have not checked it, if webpack would introduce them for some other reason...)

Reply:
`this` does not throw an error, but the next issue is `atob is not defined`. Tried to replace it with `Buffer.from`, but then `Buffer is not defined`.

Reply:
Oh thats weird, all `atob` calls should already have a guard for browser vs nodejs. Whats the actual global context in a snapshot - is this a nodejs context during snapshotting, that gets simply mapped into the browser later on? Because that the only way I can think of failing the guard.

The guard looks like this:
```typescript
function _dec(s: string): Uint8Array {
  if (typeof Buffer !== 'undefined') return Buffer.from(s, 'base64');
  const bs = atob(s);
  const r = new Uint8Array(bs.length);
  for (let i = 0; i < r.length; ++i) r[i] = bs.charCodeAt(i);
  return r;
}
```

Reply:
The issue is created by `mk-snapshot.js` in line 28:
https://github.com/vercel/hyper/blob/11c923842a997ef14ca5ad3eac0ac4cd29cf189f/bin/mk-snapshot.js#L28

Idk what `result.snapshotScript` contains at that time (some electronLink output), but the issue actually might be, that `vm.runInNewContext` creates just a naked V8 context, and not a browser typical context? Thus missing `self` and `self.atob`?

Maybe its already enough to stub `self` and `atob` in that context to pass the test?

Reply:
The v8 snapshot is just parsing the js and storing it's in memory representation, to speed up initial start. Since it doesn't need to execute it fully while snapshot creation, it just should be in the bundle, it works if we just put it in the `if(false)` block in the snapshot script. Most of the xterm packages have similar issues, you can see almost all of them are in that block.

Reply:
The following fixes `self` and `atob` not being defined:
```javascript
const vm = require('vm');
vm.runInNewContext('self.atob("AAAA")', {
    atob: s => Buffer.from(s, 'base64').toString('binary'),
    self: global
});
// --> '\x00\x00\x00'
```

Reply:
> Most of the xterm packages have similar issues, you can see almost all of them are in that block.

The errors for all those package are most likely also because of the empty global context here. Maybe its a good idea, to at least shim most often used browser API stuff in that test context?

Reply:
Will have to check how electron-link and mksnapshot are doing things to figure out where/how to modify the context. I'll take a look.

Reply:
Ah, somehow didn't register your earlier comment, will try that out.

Reply:
I tried adding the missing definitions to the context. The `vm.runInNewContext` runs but `mksnapshot` fails later on. Will need further research, I guess.